JPA: Using generic @Embeddable in Entities produces io.quarkus.builder.BuildException

### Describe the bug

When using an `@Embedded` field of a generic `@Embeddable` class, Quarkus / Hibernate now throws an error.

### Expected behavior

No error

### Actual behavior

```
java.lang.RuntimeException:
java.lang.RuntimeException: io.quarkus.builder.BuildException: Build failure: Build failed due to errors
	[error]: Build step io.quarkus.hibernate.orm.deployment.HibernateOrmProcessor#defineJpaEntities threw an exception: java.lang.IllegalStateException: io.quarkus.builder.BuildException: Build failure: root.StatusEnum is used as an embeddable but does not have an @Embeddable annotation.
	at io.quarkus.deployment.ExtensionLoader$3.execute(ExtensionLoader.java:867)
	at io.quarkus.builder.BuildContext.run(BuildContext.java:282)
	at org.jboss.threads.ContextHandler$1.runWith(ContextHandler.java:18)
	at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2513)
	at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1538)
	at java.base/java.lang.Thread.run(Thread.java:833)
	at org.jboss.threads.JBossThread.run(JBossThread.java:501)
Caused by: io.quarkus.builder.BuildException: Build failure: root.StatusEnum is used as an embeddable but does not have an @Embeddable annotation.
	at io.quarkus.hibernate.orm.deployment.JpaJandexScavenger.validateEmbeddable(JpaJandexScavenger.java:510)
	at io.quarkus.hibernate.orm.deployment.JpaJandexScavenger.collectEmbeddedTypes(JpaJandexScavenger.java:489)
	at io.quarkus.hibernate.orm.deployment.JpaJandexScavenger.collectEmbeddedTypes(JpaJandexScavenger.java:495)
	at io.quarkus.hibernate.orm.deployment.JpaJandexScavenger.enlistEmbeddedsAndElementCollections(JpaJandexScavenger.java:323)
	at io.quarkus.hibernate.orm.deployment.JpaJandexScavenger.discoverModelAndRegisterForReflection(JpaJandexScavenger.java:96)
	at io.quarkus.hibernate.orm.deployment.HibernateOrmProcessor.defineJpaEntities(HibernateOrmProcessor.java:409)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at io.quarkus.deployment.ExtensionLoader$3.execute(ExtensionLoader.java:858)
	... 6 more
```

### How to Reproduce?

1. clone https://github.com/mensinda/quarkus-stuff/tree/embeddedError
2. checkout the `embeddedError` branch
3. run `mvn clean verify`

### Output of `uname -a` or `ver`

Linux XXXXXX 5.15.0-56-generic #62-Ubuntu SMP Tue Nov 22 19:54:14 UTC 2022 x86_64 x86_64 x86_64 GNU/Linux

### Output of `java -version`

openjdk 17.0.8.1 2023-08-24

### GraalVM version (if different from Java)

_No response_

### Quarkus version or git rev

3.4.1

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

Apache Maven 3.9.2 (c9616018c7a021c1c39be70fb2843d6f5f9b8a1c)

### Additional information

Code:

```java
public enum StatusEnum implements StatusInterface {
    OK(1), WARNING(2), ERROR(3);
    // ...
}

@Embeddable
public class StatusEmbeddable<T extends StatusInterface> {

    public int code;
    public String name;

    // ...
}

@Entity
public class StatusStorageEntity {

    @Id
    long id;

    // This does no longer work!
    @Embedded
    private StatusEmbeddable<StatusEnum> embeddedStatus;
}
```

--- a/extensions/hibernate-orm/deployment/src/main/java/io/quarkus/hibernate/orm/deployment/JpaJandexScavenger.java
+++ b/extensions/hibernate-orm/deployment/src/main/java/io/quarkus/hibernate/orm/deployment/JpaJandexScavenger.java
@@ -59,8 +59,7 @@ import io.quarkus.runtime.configuration.ConfigurationException;
  */
 public final class JpaJandexScavenger {
 
-    public static final List<DotName> EMBEDDED_ANNOTATIONS = Arrays.asList(ClassNames.EMBEDDED_ID, ClassNames.EMBEDDED,
-            ClassNames.ELEMENT_COLLECTION);
+    public static final List<DotName> EMBEDDED_ANNOTATIONS = Arrays.asList(ClassNames.EMBEDDED_ID, ClassNames.EMBEDDED);
 
     private static final String XML_MAPPING_DEFAULT_ORM_XML = "META-INF/orm.xml";
     private static final String XML_MAPPING_NO_FILE = "no-file";
@@ -313,17 +312,15 @@ public final class JpaJandexScavenger {
         Set<DotName> embeddedTypes = new HashSet<>();
 
         for (DotName embeddedAnnotation : EMBEDDED_ANNOTATIONS) {
-            Collection<AnnotationInstance> annotations = index.getAnnotations(embeddedAnnotation);
-
-            for (AnnotationInstance annotation : annotations) {
+            for (AnnotationInstance annotation : index.getAnnotations(embeddedAnnotation)) {
                 AnnotationTarget target = annotation.target();
 
                 switch (target.kind()) {
                     case FIELD:
-                        collectEmbeddedTypes(embeddedAnnotation, embeddedTypes, target.asField().type());
+                        collectEmbeddedType(embeddedTypes, target.asField().type(), true);
                         break;
                     case METHOD:
-                        collectEmbeddedTypes(embeddedAnnotation, embeddedTypes, target.asMethod().returnType());
+                        collectEmbeddedType(embeddedTypes, target.asMethod().returnType(), true);
                         break;
                     default:
                         throw new IllegalStateException(
@@ -331,6 +328,23 @@ public final class JpaJandexScavenger {
                 }
 
             }
+        }
+
+        for (AnnotationInstance annotation : index.getAnnotations(ClassNames.ELEMENT_COLLECTION)) {
+            AnnotationTarget target = annotation.target();
+
+            switch (target.kind()) {
+                case FIELD:
+                    collectElementCollectionTypes(embeddedTypes, target.asField().type());
+                    break;
+                case METHOD:
+                    collectElementCollectionTypes(embeddedTypes, target.asMethod().returnType());
+                    break;
+                default:
+                    throw new IllegalStateException(
+                            "[internal error] " + ClassNames.ELEMENT_COLLECTION + " placed on a unknown element: " + target);
+            }
+
         }
 
         for (DotName embeddedType : embeddedTypes) {
@@ -481,34 +495,48 @@ public final class JpaJandexScavenger {
         }
     }
 
-    private void collectEmbeddedTypes(DotName embeddedAnnotation, Set<DotName> embeddedTypes, Type indexType)
+    private void collectEmbeddedType(Set<DotName> embeddedTypes, Type embeddedType, boolean validate)
+            throws BuildException {
+        DotName className;
+        switch (embeddedType.kind()) {
+            case CLASS:
+                className = embeddedType.asClassType().name();
+                break;
+            case PARAMETERIZED_TYPE:
+                className = embeddedType.name();
+                break;
+            default:
+                // do nothing
+                return;
+        }
+        if (validate && !index.getClassByName(className).hasAnnotation(ClassNames.EMBEDDABLE)) {
+            throw new BuildException(
+                    className + " is used as an embeddable but does not have an @Embeddable annotation.");
+        }
+        embeddedTypes.add(embeddedType.name());
+    }
+
+    private void collectElementCollectionTypes(Set<DotName> embeddedTypes, Type indexType)
             throws BuildException {
         switch (indexType.kind()) {
             case CLASS:
-                DotName className = indexType.asClassType().name();
-                validateEmbeddable(embeddedAnnotation, className);
-                embeddedTypes.add(className);
+                // Raw collection type, nothing we can do
                 break;
             case PARAMETERIZED_TYPE:
                 embeddedTypes.add(indexType.name());
-                for (Type typeArgument : indexType.asParameterizedType().arguments()) {
-                    collectEmbeddedTypes(embeddedAnnotation, embeddedTypes, typeArgument);
+                var typeArguments = indexType.asParameterizedType().arguments();
+                for (Type typeArgument : typeArguments) {
+                    // We don't validate @Embeddable annotations on element collections at the moment
+                    // See https://github.com/quarkusio/quarkus/pull/35822
+                    collectEmbeddedType(embeddedTypes, typeArgument, false);
                 }
                 break;
             case ARRAY:
-                collectEmbeddedTypes(embeddedAnnotation, embeddedTypes, indexType.asArrayType().constituent());
+                collectEmbeddedType(embeddedTypes, indexType.asArrayType().constituent(), true);
                 break;
             default:
                 // do nothing
                 break;
-        }
-    }
-
-    private void validateEmbeddable(DotName embeddedAnnotation, DotName className) throws BuildException {
-        if ((ClassNames.EMBEDDED.equals(embeddedAnnotation) || ClassNames.EMBEDDED_ID.equals(embeddedAnnotation))
-                && !index.getClassByName(className).hasAnnotation(ClassNames.EMBEDDABLE)) {
-            throw new BuildException(
-                    className + " is used as an embeddable but does not have an @Embeddable annotation.");
         }
     }
 

--- a/extensions/hibernate-orm/deployment/src/test/java/io/quarkus/hibernate/orm/enhancer/HibernateEntityEnhancerPresentEmbeddableTest.java
+++ b/extensions/hibernate-orm/deployment/src/test/java/io/quarkus/hibernate/orm/enhancer/HibernateEntityEnhancerPresentEmbeddableTest.java
@@ -1,0 +1,293 @@
+package io.quarkus.hibernate.orm.enhancer;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+import java.util.List;
+import java.util.Map;
+import java.util.SortedMap;
+import java.util.TreeMap;
+
+import jakarta.inject.Inject;
+import jakarta.persistence.ElementCollection;
+import jakarta.persistence.Embeddable;
+import jakarta.persistence.Embedded;
+import jakarta.persistence.Entity;
+import jakarta.persistence.EntityManager;
+import jakarta.persistence.GeneratedValue;
+import jakarta.persistence.Id;
+import jakarta.persistence.MappedSuperclass;
+import jakarta.persistence.OrderColumn;
+
+import org.assertj.core.api.InstanceOfAssertFactories;
+import org.hibernate.annotations.SortNatural;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.RegisterExtension;
+
+import io.quarkus.hibernate.orm.TransactionTestUtils;
+import io.quarkus.narayana.jta.QuarkusTransaction;
+import io.quarkus.test.QuarkusUnitTest;
+
+/**
+ * Checks that the missing @Embeddable check doesn't mistakely report
+ * types that are annotated with @Embeddable (https://github.com/quarkusio/quarkus/issues/35598)
+ * or generic type parameters on @Embedded field types (https://github.com/quarkusio/quarkus/issues/36065)
+ */
+public class HibernateEntityEnhancerPresentEmbeddableTest {
+
+    @RegisterExtension
+    static QuarkusUnitTest runner = new QuarkusUnitTest()
+            .withApplicationRoot((jar) -> jar
+                    .addClass(TransactionTestUtils.class)
+                    .addClasses(EntityWithEmbedded.class, EmbeddableWithAnnotation.class,
+                            ExtendedEmbeddableWithAnnotation.class,
+                            NestingEmbeddableWithAnnotation.class,
+                            GenericEmbeddableWithAnnotation.class))
+            .withConfigurationResource("application.properties")
+            .overrideConfigKey("quarkus.hibernate-orm.implicit-naming-strategy", "component-path");
+
+    @Inject
+    EntityManager em;
+
+    // Just test that the generic embeddeds work correctly over a persist/retrieve cycle
+    @Test
+    public void smokeTest() {
+        Long id = QuarkusTransaction.requiringNew().call(() -> {
+            EntityWithEmbedded entity = new EntityWithEmbedded();
+            entity.setName("name");
+            entity.setEmbeddedWithAnnotation(new EmbeddableWithAnnotation("simple"));
+            entity.setExtendedEmbeddedWithAnnotation(new ExtendedEmbeddableWithAnnotation("extended", 42));
+            var nesting = new NestingEmbeddableWithAnnotation("nesting");
+            entity.setNestingEmbeddedWithAnnotation(nesting);
+            nesting.setEmbedded(new EmbeddableWithAnnotation("nested"));
+            entity.setGenericEmbeddedWithAnnotation(new GenericEmbeddableWithAnnotation<>("generic"));
+            entity.setEmbeddableListWithAnnotation(List.of(
+                    new EmbeddableWithAnnotation("list1"),
+                    new EmbeddableWithAnnotation("list2")));
+            entity.setEmbeddableMapValueWithAnnotation(new TreeMap<>(Map.of(
+                    "first", new EmbeddableWithAnnotation("map1"),
+                    "second", new EmbeddableWithAnnotation("map2"))));
+            em.persist(entity);
+            return entity.getId();
+        });
+
+        QuarkusTransaction.requiringNew().run(() -> {
+            EntityWithEmbedded entity = em.find(EntityWithEmbedded.class, id);
+            assertThat(entity).extracting(e -> e.getName())
+                    .isEqualTo("name");
+            assertThat(entity).extracting(e -> e.getEmbeddedWithAnnotation().getText())
+                    .isEqualTo("simple");
+            assertThat(entity).extracting(e -> e.getExtendedEmbeddedWithAnnotation().getText())
+                    .isEqualTo("extended");
+            assertThat(entity).extracting(e -> e.getExtendedEmbeddedWithAnnotation().getInteger())
+                    .isEqualTo(42);
+            assertThat(entity).extracting(e -> e.getNestingEmbeddedWithAnnotation().getText())
+                    .isEqualTo("nesting");
+            assertThat(entity).extracting(e -> e.getNestingEmbeddedWithAnnotation().getEmbedded().getText())
+                    .isEqualTo("nested");
+            assertThat(entity).extracting(e -> e.getGenericEmbeddedWithAnnotation().getValue())
+                    .isEqualTo("generic");
+            assertThat(entity).extracting(e -> e.getEmbeddableListWithAnnotation())
+                    .asInstanceOf(InstanceOfAssertFactories.list(EmbeddableWithAnnotation.class))
+                    .extracting(EmbeddableWithAnnotation::getText)
+                    .containsExactly("list1", "list2");
+            assertThat(entity).extracting(e -> e.getEmbeddableMapValueWithAnnotation())
+                    .asInstanceOf(InstanceOfAssertFactories.map(String.class, EmbeddableWithAnnotation.class))
+                    .extractingFromEntries(e -> e.getValue().getText())
+                    .containsExactly("map1", "map2");
+        });
+    }
+
+    @Entity
+    public static class EntityWithEmbedded {
+
+        @Id
+        @GeneratedValue
+        private Long id;
+
+        private String name;
+
+        @Embedded
+        private EmbeddableWithAnnotation embeddedWithAnnotation;
+
+        @Embedded
+        private ExtendedEmbeddableWithAnnotation extendedEmbeddedWithAnnotation;
+
+        @Embedded
+        private NestingEmbeddableWithAnnotation nestingEmbeddedWithAnnotation;
+
+        @Embedded
+        private GenericEmbeddableWithAnnotation<String> genericEmbeddedWithAnnotation;
+
+        @ElementCollection
+        @OrderColumn
+        private List<EmbeddableWithAnnotation> embeddableListWithAnnotation;
+
+        @ElementCollection
+        @SortNatural
+        private SortedMap<String, EmbeddableWithAnnotation> embeddableMapValueWithAnnotation;
+
+        public Long getId() {
+            return id;
+        }
+
+        public void setId(Long id) {
+            this.id = id;
+        }
+
+        public String getName() {
+            return name;
+        }
+
+        public void setName(String name) {
+            this.name = name;
+        }
+
+        public EmbeddableWithAnnotation getEmbeddedWithAnnotation() {
+            return embeddedWithAnnotation;
+        }
+
+        public void setEmbeddedWithAnnotation(EmbeddableWithAnnotation embeddedWithAnnotation) {
+            this.embeddedWithAnnotation = embeddedWithAnnotation;
+        }
+
+        public ExtendedEmbeddableWithAnnotation getExtendedEmbeddedWithAnnotation() {
+            return extendedEmbeddedWithAnnotation;
+        }
+
+        public void setExtendedEmbeddedWithAnnotation(ExtendedEmbeddableWithAnnotation extendedEmbeddedWithAnnotation) {
+            this.extendedEmbeddedWithAnnotation = extendedEmbeddedWithAnnotation;
+        }
+
+        public NestingEmbeddableWithAnnotation getNestingEmbeddedWithAnnotation() {
+            return nestingEmbeddedWithAnnotation;
+        }
+
+        public void setNestingEmbeddedWithAnnotation(NestingEmbeddableWithAnnotation nestingEmbeddedWithAnnotation) {
+            this.nestingEmbeddedWithAnnotation = nestingEmbeddedWithAnnotation;
+        }
+
+        public GenericEmbeddableWithAnnotation<String> getGenericEmbeddedWithAnnotation() {
+            return genericEmbeddedWithAnnotation;
+        }
+
+        public void setGenericEmbeddedWithAnnotation(GenericEmbeddableWithAnnotation<String> genericEmbeddedWithAnnotation) {
+            this.genericEmbeddedWithAnnotation = genericEmbeddedWithAnnotation;
+        }
+
+        public List<EmbeddableWithAnnotation> getEmbeddableListWithAnnotation() {
+            return embeddableListWithAnnotation;
+        }
+
+        public void setEmbeddableListWithAnnotation(List<EmbeddableWithAnnotation> embeddableListWithAnnotation) {
+            this.embeddableListWithAnnotation = embeddableListWithAnnotation;
+        }
+
+        public Map<String, EmbeddableWithAnnotation> getEmbeddableMapValueWithAnnotation() {
+            return embeddableMapValueWithAnnotation;
+        }
+
+        public void setEmbeddableMapValueWithAnnotation(
+                SortedMap<String, EmbeddableWithAnnotation> embeddableMapValueWithAnnotation) {
+            this.embeddableMapValueWithAnnotation = embeddableMapValueWithAnnotation;
+        }
+    }
+
+    @Embeddable
+    @MappedSuperclass
+    public static class EmbeddableWithAnnotation {
+        private String text;
+
+        protected EmbeddableWithAnnotation() {
+            // For Hibernate ORM only - it will change the property value through reflection
+        }
+
+        public EmbeddableWithAnnotation(String text) {
+            this.text = text;
+        }
+
+        public String getText() {
+            return text;
+        }
+
+        public void setText(String text) {
+            this.text = text;
+        }
+    }
+
+    @Embeddable
+    public static class NestingEmbeddableWithAnnotation {
+        private String text;
+
+        @Embedded
+        private EmbeddableWithAnnotation embedded;
+
+        protected NestingEmbeddableWithAnnotation() {
+            // For Hibernate ORM only - it will change the property value through reflection
+        }
+
+        public NestingEmbeddableWithAnnotation(String text) {
+            this.text = text;
+            this.embedded = new EmbeddableWithAnnotation(text);
+        }
+
+        public String getText() {
+            return text;
+        }
+
+        public void setText(String text) {
+            this.text = text;
+        }
+
+        public EmbeddableWithAnnotation getEmbedded() {
+            return embedded;
+        }
+
+        public void setEmbedded(EmbeddableWithAnnotation embedded) {
+            this.embedded = embedded;
+        }
+    }
+
+    @Embeddable
+    public static class ExtendedEmbeddableWithAnnotation extends EmbeddableWithAnnotation {
+        private Integer integer;
+
+        protected ExtendedEmbeddableWithAnnotation() {
+            // For Hibernate ORM only - it will change the property value through reflection
+        }
+
+        public ExtendedEmbeddableWithAnnotation(String text, Integer integer) {
+            super(text);
+            this.integer = integer;
+        }
+
+        public Integer getInteger() {
+            return integer;
+        }
+
+        public void setInteger(Integer integer) {
+            this.integer = integer;
+        }
+    }
+
+    @Embeddable
+    public static class GenericEmbeddableWithAnnotation<T> {
+        private T value;
+
+        protected GenericEmbeddableWithAnnotation() {
+            // For Hibernate ORM only - it will change the property value through reflection
+        }
+
+        public GenericEmbeddableWithAnnotation(T value) {
+            this.value = value;
+        }
+
+        public T getValue() {
+            return value;
+        }
+
+        public void setValue(T value) {
+            this.value = value;
+        }
+    }
+
+}

